src/track/beatutils.cpp

```diff
@@ -29,20 +31,178 @@ const int kHistogramDecimalPlaces = 2;
+    const double median = computeSampleMedian(sortedTempoList);
+    // The analyzer sometimes detect false beats that generate outliers
+    // values for the tempo so we use a median filter to remove them
+    MovingMedian filterTempo(5); // 5 is the lenght our window
+    int currentBeat = -1;
+    int lastBeatChange = 0;
+    QMap<int, double> stableTemposAndPositions;
+    stableTemposAndPositions[lastBeatChange] = median;
+    // Here we are going to track the tempo changes over the track
+    for (double tempo : tempoList) {
+        currentBeat += 1;
+        double newStableTempo = filterTempo(tempo);
+        // The analyzer has some jitter that causes a steady beat to fluctuate around
+        // the correct value so we don't consider tempos +-1 from that as changes
+        if (newStableTempo == stableTemposAndPositions.last()) {
+            continue;
+        } else if (stableTemposAndPositions.last() != tempoFrequency.lastKey() and
+                newStableTempo == (tempoFrequency.find(stableTemposAndPositions.last()) + 1).key()) {
+            continue;
+        } else if (stableTemposAndPositions.last() != tempoFrequency.firstKey() and
+                newStableTempo == (tempoFrequency.find(stableTemposAndPositions.last()) - 1).key()) {
+            continue;
+        } else {
+            // this may not be case when our median window is even, we can't use it
+            // because find will return an iterator pointing to end that we will *
+            if (tempoFrequency.contains(newStableTempo)) {
+                lastBeatChange = currentBeat - filterTempo.lag();
+                stableTemposAndPositions[lastBeatChange] = newStableTempo;
+            }
+        }
+    }
+    stableTemposAndPositions[tempoList.count()] = median;
+    return stableTemposAndPositions;
+}
+
+
+QVector<double> BeatUtils::FixBeatmap(
+        QVector<double>& rawBeats, int sampleRate, double minBpm, double maxBpm) {
+    
+    QMap<double, int> tempoFrequency;
+    QList<double> tempoList = computeWindowedBpmsAndFrequencyHistogram(
+            rawBeats, 2, 1, sampleRate, &tempoFrequency);
+    QMap<int, double> stableTemposAndPositions = BeatUtils::findTempoChanges(
+        tempoFrequency, tempoList);
+    QVector<double> fixedBeats;
+    // We only care about where the changes happened
+    auto tempoChanges = stableTemposAndPositions.keys();    
+    for (int lastTempoChage = 0; 
+                lastTempoChage < tempoChanges.count() -1;
+                lastTempoChage++) {
+
+        int beatStart = tempoChanges[lastTempoChage];
+        int beatEnd = tempoChanges[lastTempoChage + 1];
+        auto splittedAtTempoChange = QVector<double>::fromStdVector(std::vector<double>(
+                rawBeats.begin() + beatStart, rawBeats.begin() + beatEnd));
+        double bpm = calculateBpm(splittedAtTempoChange, sampleRate, minBpm, maxBpm);
+        fixedBeats << calculateFixedTempoBeatMap(splittedAtTempoChange, sampleRate, bpm);
```

Comment:
Cool, Can we force a minimum fixed region of one measure? Can we align the change to downbeats? 
That makes musically pretty much sense. But do we have the info?

Reply:
That's for #2877, we dont have that info here